lib/src/git.rs

```diff
@@ -623,6 +625,29 @@ impl<'a> RemoteCallbacks<'a> {
                 true
             });
         }
+        if let Some(certificate_check_cb) = self.check_certificate {
+            callbacks.certificate_check(move |cert, hostname| {
+                let cert_descr: Cow<'_, str> = if let Some(hostkey) = cert.as_hostkey() {
```

Comment:
Appears that this callback is invoked no matter if the hostkey could be verified or not.

https://github.com/libgit2/libgit2/blob/e6325351ceee58cf56f58bdce61b38907805544f/src/libgit2/transports/ssh.c#L748

```
Thread 1 "jj" hit Breakpoint 1, jujutsu::commands::git::terminal_check_unknown_certificate_ok (ui=0x7fffffffca88, cert=..., url=...) at src/commands/git.rs:539
539	   ui.prompt_yes_no(&format!(
(gdb) bt
#0  jujutsu::commands::git::terminal_check_unknown_certificate_ok (ui=0x7fffffffca88, cert=..., url=...) at src/commands/git.rs:539
#1  0x0000555555e3f262 in jujutsu::commands::git::with_remote_callbacks::{closure#4}<core::result::Result<core::option::Option<alloc::string::String>, jujutsu_lib::git::GitFetchError>, jujutsu::commands::git::cmd_git_fetch::{closure_env#1}> (cert_str=..., hostname=...) at src/commands/git.rs:524
#2  0x00005555561e3eee in jujutsu_lib::git::{impl#0}::into_git::{closure#1} (cert=0x7fffffff4b40, hostname=...) at lib/src/git.rs:644
#3  0x0000555556ad1e2e in alloc::boxed::{impl#46}::call_mut<(&git2::cert::Cert, &str), dyn core::ops::function::FnMut<(&git2::cert::Cert, &str), Output=core::result::Result<git2::remote_callbacks::CertificateCheckStatus, git2::error::Error>>, alloc::alloc::Global> (self=0x7fffffff60b0, args=...)
    at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/alloc/src/boxed.rs:1994
#4  0x0000555556ad884d in git2::remote_callbacks::certificate_check_cb::{closure#0} () at src/remote_callbacks.rs:413
#5  0x0000555556ae4831 in std::panicking::try::do_call<git2::remote_callbacks::certificate_check_cb::{closure_env#0}, core::result::Result<git2::remote_callbacks::CertificateCheckStatus, git2::error::Error>> (data=0x7fffffff4cb0) at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:487
#6  0x0000555556ae558b in __rust_try ()
#7  0x0000555556ae3d1d in std::panicking::try<core::result::Result<git2::remote_callbacks::CertificateCheckStatus, git2::error::Error>, git2::remote_callbacks::certificate_check_cb::{closure_env#0}> (f=...) at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:451
#8  0x0000555556ad130e in std::panic::catch_unwind<git2::remote_callbacks::certificate_check_cb::{closure_env#0}, core::result::Result<git2::remote_callbacks::CertificateCheckStatus, git2::error::Error>> (f=...) at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panic.rs:140
#9  0x0000555556addb40 in git2::panic::wrap<core::result::Result<git2::remote_callbacks::CertificateCheckStatus, git2::error::Error>, git2::remote_callbacks::certificate_check_cb::{closure_env#0}> (f=...) at src/panic.rs:13
#10 0x0000555556ad8517 in git2::remote_callbacks::certificate_check_cb (cert=0x7fffffff4fb0, _valid=1, hostname=0x5555571406c0, data=0x7fffffff6050)
    at src/remote_callbacks.rs:405
#11 0x0000555556bdb5e0 in check_certificate (session=0x5555574923c0, known_hosts=0x5555570a9960, check_cb=0x555556ad84c0 <git2::remote_callbacks::certificate_check_cb>, 
    check_cb_payload=0x7fffffff6050, host=0x5555571406c0 "github.com", port=22) at libgit2/src/libgit2/transports/ssh.c:753
#12 0x0000555556bdb822 in _git_ssh_setup_conn (t=0x5555571406e0, url=0x55555716d960 "git@github.com:yuja/jj.git", cmd=0x555555b8ea00 <cmd_uploadpack> "git-upload-pack", 
    stream=0x7fffffff5178) at libgit2/src/libgit2/transports/ssh.c:811
#13 0x0000555556bdbbb7 in ssh_uploadpack_ls (t=0x5555571406e0, url=0x55555716d960 "git@github.com:yuja/jj.git", stream=0x7fffffff5178)
    at libgit2/src/libgit2/transports/ssh.c:903
#14 0x0000555556bdbd29 in _ssh_action (stream=0x7fffffff5178, subtransport=0x5555571406e0, url=0x55555716d960 "git@github.com:yuja/jj.git", 
    action=GIT_SERVICE_UPLOADPACK_LS) at libgit2/src/libgit2/transports/ssh.c:959
#15 0x0000555556bd4e45 in git_smart__connect (transport=0x55555714e040, url=0x5555571f6c50 "git@github.com:yuja/jj.git", direction=0, connect_opts=0x7fffffff5200)
    at libgit2/src/libgit2/transports/smart.c:152
#16 0x0000555556b948e5 in git_remote_connect_ext (remote=0x55555712b420, direction=GIT_DIRECTION_FETCH, given_opts=0x7fffffff5330) at libgit2/src/libgit2/remote.c:962
#17 0x0000555556b952a3 in connect_or_reset_options (remote=0x55555712b420, direction=0, opts=0x7fffffff5330) at libgit2/src/libgit2/remote.c:1253
#18 0x0000555556b95623 in git_remote_download (remote=0x55555712b420, refspecs=0x7fffffff54b0, opts=0x7fffffff55b0) at libgit2/src/libgit2/remote.c:1338
#19 0x000055555623cd0d in git2::remote::Remote::download<alloc::string::String> (self=0x7fffffff5f48, specs=..., opts=...)
    at /home/yuya/opt/rustup/registry/src/github.com-1ecc6299db9ec823/git2-0.17.1/src/remote.rs:249
#20 0x0000555556320a8b in jujutsu_lib::git::fetch (mut_repo=0x7fffffff94c8, git_repo=0x7fffffff9338, remote_name=..., branch_name_globs=..., callbacks=..., 
    git_settings=0x7fffffff7a77) at lib/src/git.rs:407
#21 0x0000555555e4018a in jujutsu::commands::git::cmd_git_fetch::{closure#1} (cb=...) at src/commands/git.rs:308
#22 0x0000555555e3d2f2 in jujutsu::commands::git::with_remote_callbacks<core::result::Result<core::option::Option<alloc::string::String>, jujutsu_lib::git::GitFetchError>, jujutsu::commands::git::cmd_git_fetch::{closure_env#1}> (ui=0x7fffffffca88, f=...) at src/commands/git.rs:527
#23 0x0000555555c07894 in jujutsu::commands::git::cmd_git_fetch (ui=0x7fffffffca88, command=0x7fffffffb028, args=0x7fffffff9fa0) at src/commands/git.rs:307
#24 0x0000555555bfed21 in jujutsu::commands::git::cmd_git (ui=0x7fffffffca88, command=0x7fffffffb028, subcommand=0x7fffffff9fa0) at src/commands/git.rs:1013
#25 0x0000555555d60996 in jujutsu::commands::run_command (ui=0x7fffffffca88, command_helper=0x7fffffffb028, matches=0x7fffffffa720) at src/commands/mod.rs:3520
#26 0x0000555555dfe96f in core::ops::function::FnOnce::call_once<fn(&mut jujutsu::ui::Ui, &jujutsu::cli_util::CommandHelper, &clap::parser::matches::arg_matches::ArgMatches) -> core::result::Result<(), jujutsu::cli_util::CommandError>, (&mut jujutsu::ui::Ui, &jujutsu::cli_util::CommandHelper, &clap::parser::matches::arg_matches::ArgMatches)> () at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/ops/function.rs:250
#27 0x0000555555dfdf13 in core::ops::function::FnOnce::call_once<fn(&mut jujutsu::ui::Ui, &jujutsu::cli_util::CommandHelper, &clap::parser::matches::arg_matches::ArgMatches) -> core::result::Result<(), jujutsu::cli_util::CommandError>, (&mut jujutsu::ui::Ui, &jujutsu::cli_util::CommandHelper, &clap::parser::matches::arg_matches::ArgMatches)> () at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/ops/function.rs:250
#28 0x0000555555cf9464 in alloc::boxed::{impl#45}::call_once<(&mut jujutsu::ui::Ui, &jujutsu::cli_util::CommandHelper, &clap::parser::matches::arg_matches::ArgMatches), dyn core::ops::function::FnOnce<(&mut jujutsu::ui::Ui, &jujutsu::cli_util::CommandHelper, &clap::parser::matches::arg_matches::ArgMatches), Output=core::result::Result<(), jujutsu::cli_util::CommandError>>, alloc::alloc::Global> (self=..., args=...) at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/alloc/src/boxed.rs:1987
#29 0x0000555555dc0d25 in jujutsu::cli_util::CliRunner::run_internal (self=..., ui=0x7fffffffca88, layered_configs=...) at src/cli_util.rs:2383
#30 0x0000555555dc1191 in jujutsu::cli_util::CliRunner::run (self=...) at src/cli_util.rs:2391
#31 0x0000555555bc18ef in jj::main () at src/main.rs:18
(gdb) frame 11
#11 0x0000555556bdb5e0 in check_certificate (session=0x5555574923c0, known_hosts=0x5555570a9960, check_cb=0x555556ad84c0 <git2::remote_callbacks::certificate_check_cb>, 
    check_cb_payload=0x7fffffff6050, host=0x5555571406c0 "github.com", port=22) at libgit2/src/libgit2/transports/ssh.c:753
753			error = check_cb((git_cert *) cert_ptr, cert_valid, host, check_cb_payload);
(gdb) info locals
cert_ptr = 0x7fffffff4fb0
previous_error = {error_code = 0, oom = 0, error_msg = {message = 0x0, klass = 0}}
cert = {parent = {cert_type = GIT_CERT_HOSTKEY_LIBSSH2}, type = (GIT_CERT_SSH_MD5 | GIT_CERT_SSH_SHA1 | GIT_CERT_SSH_SHA256 | GIT_CERT_SSH_RAW), 
  hash_md5 = "e\226-\374\350թ\021d\f\017\352\000n[\275", hash_sha1 = "\351a\236.\325l/*qr\235\270\v\254\302Μ\314\350", <incomplete sequence \324>, 
  hash_sha256 = "\3708\230\337\v\357W\244\356$\230[\245\230\254\027\374\313\f\r3<į\035\331+\341K\302:\245", raw_type = GIT_CERT_SSH_RAW_TYPE_KEY_ED25519, 
  hostkey = 0x5555574be310 "", hostkey_len = 51}
key = 0x55555749247c "e\226-\374\350թ\021d\f", <incomplete sequence \352>
cert_len = 51
cert_type = 6
cert_valid = 1
error = 0
(
```

Reply:
Ah... The doc for `certificate_check` (which seems to be copied from the [libgit2 docs](https://libgit2.org/libgit2/#HEAD/type/git_remote_callbacks)) says

> If certificate verification fails, then this callback will be invoked to
> let the caller make the final decision [...]

So I assumed this would be fine. However, the documentation for `CertificateCheckStatus::CertificatePassthrough` says

> Indicates that the certificate callback is neither accepting nor
> rejecting the certificate. The result of the certificate checks
> built-in to libgit2 will be used instead.

Which seems to back up the behavior you're describing.

I'll try to find an alternative solution to `certificate_check`.

Reply:
So:
1. A custom implementation of the check is described in this discussion: https://github.com/libgit2/libgit2/discussions/5928. It may be useful if we need to implement it.
2. It looks like `libgit2` has a different behavior for `certificate_check` on Windows: https://github.com/libgit2/libgit2/blob/8a62616f43fe5ea37d41296f40293ff97aa88cfa/src/libgit2/transports/winhttp.c#L296. If we could get it to use the [SSH transport](https://github.com/libgit2/libgit2/blob/8a62616f43fe5ea37d41296f40293ff97aa88cfa/src/libgit2/transports/ssh.c#L811) (which uses libssh2), this would likely work better.

Reply:
Resolving as this change is no longer needed.